File: reconstruction/CaloRec/src/PulseGenerator.cxx
#include "PulseGenerator.h"
#include "TPulseGenerator.h"

using namespace Gaugi;


PulseGenerator::PulseGenerator( std::string name ) : 
  IMsgService(name),
  CaloTool()
{
  declareProperty( "NSamples"     , m_nsamples=7            );
  declareProperty( "ShaperFile"   , m_shaperFile            );
  declareProperty( "OutputLevel"  , m_outputLevel=1         );
}



PulseGenerator::~PulseGenerator()
{
  delete m_pulseGenerator;
}


StatusCode PulseGenerator::initialize()
{
  setMsgLevel( (MSG::Level)m_outputLevel );
  MSG_DEBUG( "Reading shaper values from: " << m_shaperFile );
  m_pulseGenerator = new CPK::TPulseGenerator( m_nsamples, m_shaperFile.c_str());
  return StatusCode::SUCCESS;
}


StatusCode PulseGenerator::finalize()
{
  return StatusCode::SUCCESS;
}


StatusCode PulseGenerator::executeTool( const xAOD::EventInfo * /*evt*/, xAOD::RawCell *cell ) const
{
  auto pulse_size = m_pulseGenerator->GetPulseSize();
  
  // Get all energies for each bunch crossing 
  auto rawEnergySamples = cell->rawEnergySamples();

  // Create an pulse with zeros with n samples
  std::vector<float> pulse_sum(pulse_size, 0.0);
  // Loop over each bunch crossing
  for ( int bc = cell->bcid_start(), i=0;  bc <= cell->bcid_end(); ++bc, ++i )
  {
    // Generate the pulse
    auto pulse = m_pulseGenerator->GenerateDeterministicPulse( rawEnergySamples[i], 0, bc*cell->bc_duration() );
    // Add gaussian noise
    // m_pulseGenerator[cell->layer()]->AddGaussianNoise(pulse);
    // Accumulate into pulse sum (Sum all pulses)
    for ( int j=0; j < pulse_size; ++j )
      pulse_sum[j] += (float)pulse->operator[](j);
    delete pulse; // This must be deleted to avoid memory leak since spk uses "new" internally
  }


  // Add the pulse centered in the bunch crossing zero
  cell->setPulse( pulse_sum );
  return StatusCode::SUCCESS;
}



// Just for python import in ROOT
StatusCode PulseGenerator::executeTool( const xAOD::EventInfo *, xAOD::CaloCell * ) const {return StatusCode::SUCCESS;}
StatusCode PulseGenerator::executeTool( const xAOD::EventInfo *, xAOD::CaloCluster * ) const {return StatusCode::SUCCESS;}
StatusCode PulseGenerator::executeTool( const xAOD::EventInfo *, xAOD::TruthParticle * ) const {return StatusCode::SUCCESS;}

File: reconstruction/CaloRec/python/PulseGenerator.py

__all__ = ["PulseGenerator"]

from Gaugi import Logger
from Gaugi.messenger.macros import *


class PulseGenerator( Logger ):

  __allow_keys = ["OutputLevel", "NSamples", "ShaperFile"]


  def __init__( self, name, **kw ):

    Logger.__init__(self)
    import ROOT
    ROOT.gSystem.Load('liblorenzetti')
    from ROOT import RunManager, PulseGenerator
    self.__core = PulseGenerator(name)
    for key, value in kw.items():
      if key in self.__allow_keys:
        setattr( self, '__' + key , value )
        self.__core.setProperty( key, value )
      else:
        MSG_ERROR( self, "Property with name %s is not allow for PulseGenerator object", key)


  def core(self):
    print (type(self.__core))
    return self.__core


  def setProperty( self, key, value ):
    if key in self.__allow_keys:
      self.core().setProperty( key, value )
    else:
      MSG_ERROR( self, "Property with name %s is not allow for PulseGenerator object")

 
  def getProperty( self, key ):
    if key in self.__allow_keys:
      return getattr( self, '__' + key )
    else:
      MSG_ERROR( self, "Property with name %s is not allow for PulseGenerator object")

     

